src/backend/distributed/transaction/multi_shard_transaction.c

```diff
@@ -43,18 +43,95 @@ void
 OpenTransactionsToAllShardPlacements(List *shardIntervalList, char *userName)
 {
 	ListCell *shardIntervalCell = NULL;
+	List *newConnectionList = NIL;
+	ListCell *connectionCell = NULL;
 
 	if (shardConnectionHash == NULL)
 	{
 		shardConnectionHash = CreateShardConnectionHash(TopTransactionContext);
 	}
 
+	BeginOrContinueCoordinatedTransaction();
+	if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC)
+	{
+		CoordinatedTransactionUse2PC();
+	}
+
+	/* open connections to shards which don't have connections yet */
 	foreach(shardIntervalCell, shardIntervalList)
 	{
 		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
 		uint64 shardId = shardInterval->shardId;
+		ShardConnections *shardConnections = NULL;
+		bool shardConnectionsFound = false;
+		List *shardPlacementList = NIL;
+		ListCell *placementCell = NULL;
+
+		shardConnections = GetShardConnections(shardId, &shardConnectionsFound);
+		if (shardConnectionsFound)
+		{
+			continue;
+		}
+
+		shardPlacementList = FinalizedShardPlacementList(shardId);
+		if (shardPlacementList == NIL)
+		{
+			/* going to have to have some placements to do any work */
+			ereport(ERROR, (errmsg("could not find any shard placements for the shard "
+								   UINT64_FORMAT, shardId)));
+		}
 
-		BeginTransactionOnShardPlacements(shardId, userName);
+		foreach(placementCell, shardPlacementList)
+		{
+			ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(placementCell);
+			MultiConnection *connection = NULL;
+			MemoryContext oldContext = NULL;
+
+			WorkerNode *workerNode = FindWorkerNode(shardPlacement->nodeName,
+													shardPlacement->nodePort);
+			if (workerNode == NULL)
+			{
+				ereport(ERROR, (errmsg("could not find worker node %s:%d",
+									   shardPlacement->nodeName,
+									   shardPlacement->nodePort)));
+			}
+
+			connection = StartNodeUserDatabaseConnection(FORCE_NEW_CONNECTION,
+														 shardPlacement->nodeName,
+														 shardPlacement->nodePort,
+														 userName,
+														 NULL);
+
+			/* we need to preserve the connection list for the next statement */
+			oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+			shardConnections->connectionList = lappend(shardConnections->connectionList,
+													   connection);
+
+			MemoryContextSwitchTo(oldContext);
+
+			newConnectionList = lappend(newConnectionList, connection);
+
+			/*
+			 * Every individual failure should cause entire distributed
+			 * transaction to fail.
+			 */
+			MarkRemoteTransactionCritical(connection);
+		}
+	}
+
+	/* finish connetion establishment newly opened connections */
+	foreach(connectionCell, newConnectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+
+		FinishConnectionEstablishment(connection);
+	}
+
+	/* the special BARE mode (for e.g. VACUUM/ANALYZE) skips BEGIN */
+	if (MultiShardCommitProtocol > COMMIT_PROTOCOL_BARE)
```

Comment:
Not a blocker, but given that the `VACUUM`/`ANALYZE` stuff piggybacks on these methods, do you see an easy way forward to avoid `COMMIT_PROTOCOL_BARE`? When connection things were still in flux, it was simple to add, but now I wonder if we couldn't add some sort of boolean here to opt-out of a coordinated transaction when running those commands…